File: indexes/python/ivf_pq.py
"""IVF with PQ-coded residuals (CONTRACT 6.5).

Recall floor (CONTRACT 9, dev set): defaults, recall@10 >= 0.45 (metric=ip and metric=l2).
"""

import numpy as np

BUILD_PARAMS: dict = {"nlist": 1024, "iters": 20, "train_size": 100000, "m": 48, "nbits": 8, "metric": "ip"}
SEARCH_PARAMS: dict = {"nprobe": 8, "rerank": 0}


def build(vectors: np.ndarray, params: dict, threads: int, seed: int) -> dict:
    raise NotImplementedError("not implemented")


def search(index: dict, query: np.ndarray, k: int, params: dict) -> tuple[np.ndarray, np.ndarray]:
    raise NotImplementedError("not implemented")


def index_bytes(index: dict) -> int:
    raise NotImplementedError("not implemented")

File: indexes/rust/src/ivf_pq.rs
//! IVF with PQ-coded residuals (CONTRACT 6.5).
//!
//! Recall floor (CONTRACT 9): recall@10 >= 0.45 at default params on the dev set, for metric=ip and metric=l2.
//!
//! Not implemented yet (Wave 2). `build` returns an error, and `bench` exits with code 1.

use crate::{AnnIndex, BuildTimes, Matrix, ParamValue::*, Params, SearchResult};

pub fn build_defaults(_n: usize) -> Params {
    Params::new()
        .with("nlist", Int(1024))
        .with("iters", Int(20))
        .with("m", Int(48))
        .with("nbits", Int(8))
        .with("metric", Str("ip".into()))
        .with("train_size", Int(100_000))
}

pub fn search_defaults() -> Params {
    Params::new().with("nprobe", Int(8)).with("rerank", Int(0))
}

pub struct IvfPqIndex {
    times: BuildTimes,
}

pub fn build(
    _vectors: Matrix,
    _params: &Params,
    _threads: usize,
    _seed: u64,
) -> Result<IvfPqIndex, String> {
    Err("not implemented".into())
}

pub fn search(
    _index: &IvfPqIndex,
    _query: &[f32],
    _k: usize,
    _params: &Params,
) -> Result<SearchResult, String> {
    Err("not implemented".into())
}

pub fn index_bytes(_index: &IvfPqIndex) -> u64 {
    0
}

impl AnnIndex for IvfPqIndex {
    fn search(&self, query: &[f32], k: usize, params: &Params) -> Result<SearchResult, String> {
        search(self, query, k, params)
    }
    fn index_bytes(&self) -> u64 {
        index_bytes(self)
    }
    fn build_times(&self) -> BuildTimes {
        self.times
    }
}
